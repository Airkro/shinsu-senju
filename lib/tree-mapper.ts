import { Table2Treed } from './table-grouping.ts';
import type { GroupNode } from './table-grouping.ts';
import { getBy } from './utils.ts';
import type { DataRecord } from './utils.ts';

export interface TreeNode {
  label: unknown;
  value: unknown;
  extra?: unknown;
  selectable?: boolean;
  children?: Tree;
  $meta?: GroupNode['$meta'];
  $original?: object;
}

export type Tree = TreeNode[];

interface Calc {
  when: string;
  const?: unknown;
  enum?: unknown[];
}

export interface Mappers {
  label?: string;
  value?: string;
  extra?: string;
  selectable?: Calc;
  disabled?: Calc;
}

function calcSelectable(node: DataRecord, matcher: Calc): undefined | boolean {
  return matcher.const
    ? getBy(node, matcher.when) === matcher.const
    : matcher.enum
      ? matcher.enum.includes(getBy(node, matcher.when))
      : Boolean(getBy(node, matcher.when));
}

export function treeMapper(data: Table2Treed, options: Mappers = {}): Tree {
  if (data.length === 0) {
    return data as Tree;
  }

  const {
    label: labelPath = 'name',
    value: valuePath = 'id',
    extra = '',
  } = options;

  return data.map((node: GroupNode | DataRecord) => {
    const children =
      node.children && Array.isArray(node.children)
        ? treeMapper(node.children, options)
        : undefined;

    if ('$meta' in node) {
      const io = node as GroupNode;

      const meta = io.$meta;

      return {
        $meta: meta,
        label: ((meta.label || meta.value) as string) || meta.groupBy,
        value: meta.value,
        ...(extra && { extra: getBy(meta, extra) }),
        ...(children && { children }),
        selectable: false,
      };
    }

    const value = getBy(node, valuePath);

    const { children: _, ...$original } = node;

    return {
      label: (getBy(node, labelPath) || value) as string,
      value,
      ...(extra && { extra: getBy(node, extra) }),
      ...(children && { children }),
      ...(options.selectable?.when && {
        selectable: calcSelectable(node, options.selectable),
      }),
      ...(options.disabled?.when && {
        disabled: calcSelectable(node, options.disabled),
      }),
      $original,
    };
  });
}
