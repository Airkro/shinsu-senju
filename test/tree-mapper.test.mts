import { expect, it } from 'vitest';

import type { Table2Treed } from '../lib/table2tree.ts';
import { treeMapper } from '../lib/tree-mapper.ts';

it('should transform group nodes correctly', () => {
  const input: Table2Treed = [
    {
      $meta: {
        groupBy: 'category',
        value: 'A',
        label: 'Category A',
      },
      children: [
        { id: 1, name: 'Item 1', category: 'A' },
        { id: 2, name: 'Item 2', category: 'A' },
      ],
    },
  ];

  expect(treeMapper(input)).toMatchSnapshot();
});

// 数据记录节点转换的快照测试
it('should transform data records correctly', () => {
  const input: Table2Treed = [{ id: 1, name: 'Test Item', category: 'X' }];

  expect(treeMapper(input)).toMatchSnapshot();
});

// 嵌套分组的快照测试
it('should handle nested groups', () => {
  const input: Table2Treed = [
    {
      $meta: {
        groupBy: 'level1',
        value: 'A',
        label: 'Level 1',
      },
      children: [
        {
          $meta: {
            groupBy: 'level2',
            value: 'X',
            label: 'Level 2',
          },
          children: [{ id: 1, name: 'Item 1', level1: 'A', level2: 'X' }],
        },
      ],
    },
  ];

  expect(treeMapper(input)).toMatchSnapshot();
});

// 空输入的快照测试
it('should handle empty input', () => {
  const input: Table2Treed = [];
  expect(treeMapper(input)).toMatchSnapshot();
});

// 缺失属性的快照测试
it('should handle missing properties', () => {
  const input: Table2Treed = [
    { id: 1 }, // 缺少name
    {
      $meta: {
        groupBy: 'category',
        value: 'B',
        // 缺少label
      },
      children: [],
    },
  ];

  expect(treeMapper(input)).toMatchSnapshot();
});

// 复杂数据结构的快照测试
it('should handle complex data structures', () => {
  const input: Table2Treed = [
    {
      $meta: {
        groupBy: 'department',
        value: 'Engineering',
        label: 'Engineering Team',
      },
      children: [
        {
          $meta: {
            groupBy: 'team',
            value: 'Frontend',
            label: 'Frontend Team',
          },
          children: [
            { id: 1, name: 'Alice', role: 'Developer', level: 'Senior' },
            { id: 2, name: 'Bob', role: 'Designer', level: 'Mid' },
          ],
        },
        {
          $meta: {
            groupBy: 'team',
            value: 'Backend',
            label: 'Backend Team',
          },
          children: [
            { id: 3, name: 'Charlie', role: 'Developer', level: 'Lead' },
            { id: 4, name: 'David', role: 'DevOps', level: 'Senior' },
          ],
        },
      ],
    },
  ];

  expect(treeMapper(input)).toMatchSnapshot();
});
