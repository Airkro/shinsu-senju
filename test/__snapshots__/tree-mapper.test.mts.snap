// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should handle complex data structures 1`] = `
[
  {
    "children": [
      {
        "children": [
          {
            "children": undefined,
            "label": "Alice",
            "value": 1,
          },
          {
            "children": undefined,
            "label": "Bob",
            "value": 2,
          },
        ],
        "label": "Frontend Team",
        "selectable": false,
        "value": "Frontend",
      },
      {
        "children": [
          {
            "children": undefined,
            "label": "Charlie",
            "value": 3,
          },
          {
            "children": undefined,
            "label": "David",
            "value": 4,
          },
        ],
        "label": "Backend Team",
        "selectable": false,
        "value": "Backend",
      },
    ],
    "label": "Engineering Team",
    "selectable": false,
    "value": "Engineering",
  },
]
`;

exports[`should handle empty input 1`] = `[]`;

exports[`should handle missing properties 1`] = `
[
  {
    "children": undefined,
    "label": undefined,
    "value": 1,
  },
  {
    "children": [],
    "label": "B",
    "selectable": false,
    "value": "B",
  },
]
`;

exports[`should handle nested groups 1`] = `
[
  {
    "children": [
      {
        "children": [
          {
            "children": undefined,
            "label": "Item 1",
            "value": 1,
          },
        ],
        "label": "Level 2",
        "selectable": false,
        "value": "X",
      },
    ],
    "label": "Level 1",
    "selectable": false,
    "value": "A",
  },
]
`;

exports[`should transform data records correctly 1`] = `
[
  {
    "children": undefined,
    "label": "Test Item",
    "value": 1,
  },
]
`;

exports[`should transform group nodes correctly 1`] = `
[
  {
    "children": [
      {
        "children": undefined,
        "label": "Item 1",
        "value": 1,
      },
      {
        "children": undefined,
        "label": "Item 2",
        "value": 2,
      },
    ],
    "label": "Category A",
    "selectable": false,
    "value": "A",
  },
]
`;
